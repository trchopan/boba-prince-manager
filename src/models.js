export const profileParser = data => ({
  fullname: data.fullname || "",
  position: data.position || "user",
  avatar: data.avatar || "/images/bloody-smile.jpg"
});
