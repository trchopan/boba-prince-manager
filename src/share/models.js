export const themes = {
  light: "Light",
  dark: "Dark"
};

export const languages = {
  vi: { code: "vi", name: "Tiáº¿ng Viá»‡t", char: "ðŸ‡»ðŸ‡³" },
  en: { code: "en", name: "English", char: "ðŸ‡¬ðŸ‡§" }
};

export const profileParser = data => ({
  fullname: data.fullname || "",
  position: data.position || "user",
  avatar: data.avatar || "/images/bloody-smile.jpg"
});
