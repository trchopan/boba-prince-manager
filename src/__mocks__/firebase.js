export const fakeAuthUser = { uid: "123456", email: "tester@arkham.city" };
export const fakeError = { code: "fake-error" };
export const docData = { name: "tester" };

const docResult = {
  exists: true,
  data: () => docData
};

const collection = jest
  .fn()
  .mockImplementationOnce(() => ({
    doc: () => ({
      get: () => Promise.resolve(docResult)
    })
  }))
  .mockImplementationOnce(() => ({
    doc: () => ({
      get: () => Promise.reject(fakeError)
    })
  }));

export const fireStore = { collection };

const signInWithEmailAndPassword = jest
  .fn()
  .mockImplementationOnce(() => Promise.resolve())
  .mockImplementationOnce(() => Promise.reject(fakeError));
const signOut = jest
  .fn()
  .mockImplementationOnce(() => Promise.resolve())
  .mockImplementationOnce(() => Promise.reject(fakeError));
export const fireAuth = { signInWithEmailAndPassword, signOut };
