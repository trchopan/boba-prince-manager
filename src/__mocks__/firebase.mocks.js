export const fakeSignUpCredentials = {
  email: "test@tester.com",
  password: "123456",
  response: "abcxyz123"
};
export const fakeAuthUser = { uid: "123456", email: "tester@arkham.city" };
export const fakeError = { code: "fake-error" };
export const docData = {
  fullname: "Tester",
  avatar: "avatar",
  position: "manager"
};

export const createSnapshot = (data) => ({
  exists: true,
  data: () => data
});

export const createFirestoreOnSnapshot = (docSnapshot) => ({
  collection: () => ({
    doc: () => ({
      onSnapshot: (cb, _) => cb(docSnapshot)
    })
  }) 
})

export const createFirestoreOnSnapshotError = (error) => ({
  collection: () => ({
    doc: () => ({
      onSnapshot: (_, err) => err(error)
    })
  }) 
})
