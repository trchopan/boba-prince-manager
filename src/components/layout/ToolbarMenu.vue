<template>
  <v-menu
    :close-on-content-click="false"
    v-model="menuOpen"
    offset-y
    left>
    <v-avatar
      slot="activator"
      :size="$vuetify.breakpoint.mdAndUp ? 42 : 36">
      <img v-if="profile" :src="profile.avatar" alt="avatar" />
      <img v-else src="/images/bloody-smile.jpg" alt="avatar" />
    </v-avatar>

    <v-card>
      <v-list v-if="profile">
        <v-list-tile avatar to="/profile" @click="menuOpen = false">
          <v-list-tile-avatar>
            <img :src="profile.avatar" alt="big avatar">
          </v-list-tile-avatar>
          <v-list-tile-content>
            <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
            <v-list-tile-sub-title>{{ profile.position | titleCase }}</v-list-tile-sub-title>
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
      <v-divider v-if="profile" />
      <v-list>
        <v-list-tile avatar>
          <v-list-tile-avatar tile>
            <img :src="`./icons/${language.code}.svg`" alt="language icon" />
          </v-list-tile-avatar>
          <v-list-tile-content>
            <v-select
              :value="language.name"
              :items="Object.values(languages).map(x => x.name)"
              :label="$t.selectLanguage"
              @change="changeLanguage($event)" />
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
      <v-card-actions>
        <v-switch
          class="pr-3 pt-2"
          :label="$t.darkTheme"
          :input-value="theme"
          :true-value="themes.dark"
          :false-value="themes.light"
          @change="changeTheme($event)">
        </v-switch>
        <v-spacer></v-spacer>
        <v-btn v-if="profile"
          outline
          @click="logout()">
          {{ $t.logout }}
        </v-btn>
        <v-btn v-else
          outline
          to="/login"
          @click="menuOpen = false">
          {{ $t.login }}
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-menu>
</template>

<script>
import { mapGetters } from "vuex";
import { themes, languages, translate } from "@/store/modules/layout";

const languagesMap = {
  selectLanguage: { vi: "Chọn ngôn ngữ", en: "Select language" },
  darkTheme: { vi: "Tắt đèn", en: "Light out" },
  logout: { vi: "Đăng xuất", en: "Log out" },
  login: { vi: "Đăng nhập", en: "Log in" }
};

export default {
  name: "ToolbarMenu",
  data: () => ({
    themes,
    languages,
    title: process.env.VUE_APP_TITLE,
    menuOpen: false
  }),
  computed: {
    ...mapGetters({
      profile: "auth/profile",
      theme: "layout/theme",
      language: "layout/language"
    }),
    $t() {
      return translate(languagesMap, this.language.code);
    }
  },
  methods: {
    async logout() {
      const loggedOut = await this.$store.dispatch("auth/logout");
      if (loggedOut) {
        this.menuOpen = false;
        this.$router.replace("/login");
      }
    },
    changeTheme(theme) {
      this.$store.dispatch("layout/changeTheme", theme);
    },
    changeLanguage(language) {
      const lang = Object.values(languages).find(x => x.name === language);
      this.$store.dispatch("layout/changeLanguage", lang);
    }
  }
};
</script>
