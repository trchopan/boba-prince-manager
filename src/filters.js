import Vue from "vue";

export const filters = {
  titleCase: str => {
    if (typeof str !== "string" || str.length === 0) {
      return "";
    }
    return str
      .toLowerCase()
      .split(" ")
      .map(
        word => (word[0] ? word.replace(word[0], word[0].toUpperCase()) : word)
      )
      .join(" ");
  }
};

Object.keys(filters).forEach(key => Vue.filter(key, filters[key]));
