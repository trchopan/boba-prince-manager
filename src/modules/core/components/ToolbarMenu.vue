<template>
  <v-menu :close-on-content-click="false" v-model="menuOpen" offset-y left>
    <v-avatar slot="activator" :size="$vuetify.breakpoint.mdAndUp ? 42 : 36">
      <img v-if="profile" :src="profile.avatar" alt="avatar">
      <img v-else :src="defaultAvatar" alt="avatar">
    </v-avatar>
    <v-card>
      <v-list v-if="authUser" two-line>
        <v-list-tile avatar to="/profile" @click="menuOpen = false">
          <v-list-tile-avatar>
            <img v-if="profile" :src="profile.avatar" alt="big avatar">
            <img v-else :src="defaultAvatar" alt="big avatar">
          </v-list-tile-avatar>
          <v-list-tile-content v-if="profile">
            <v-list-tile-title>{{ profile.fullname }}</v-list-tile-title>
            <v-list-tile-sub-title>{{ authUser.email }}</v-list-tile-sub-title>
          </v-list-tile-content>
          <v-list-tile-content v-else>
            <v-list-tile-title>{{ authUser.email }}</v-list-tile-title>
            <v-list-tile-sub-title>{{ $t.updateProfile }}</v-list-tile-sub-title>
          </v-list-tile-content>
        </v-list-tile>
      </v-list>
      <v-divider v-if="authUser"/>
      <v-list>
        <v-list-tile>
          <v-list-tile-content>
            <v-select
              :value="language"
              :items="languages"
              :label="$t.selectLanguage"
              @change="changeLanguage($event)"
            />
          </v-list-tile-content>
        </v-list-tile>
        <v-list-tile>
          <v-switch
            class="pr-3 pt-2"
            :label="$t.darkTheme"
            :input-value="theme"
            :true-value="themes.dark"
            :false-value="themes.light"
            @change="changeTheme($event)"
          ></v-switch>
        </v-list-tile>
      </v-list>
      <v-card-actions>
        <v-spacer></v-spacer>
        <v-btn v-if="authUser" outline @click="logout()">{{ $t.logout }}</v-btn>
        <v-btn
          v-else
          outline
          to="/login"
          @click="menuOpen = false"
        >{{ $t.login }}</v-btn>
      </v-card-actions>
    </v-card>
  </v-menu>
</template>

<script>
import { mapGetters } from "vuex";
import { themes, languages } from "@/modules/core/root.models.js";

export const LANGUAGES_MAP = {
  updateProfile: {
    vi: "Cập nhật thông tin tài khoản",
    en: "Please update your profile"
  },
  selectLanguage: { vi: "Chọn ngôn ngữ", en: "Select language" },
  darkTheme: { vi: "Tắt đèn", en: "Light out" },
  logout: { vi: "Đăng xuất", en: "Log out" },
  login: { vi: "Đăng nhập", en: "Log in" }
};

export default {
  name: "ToolbarMenu",
  data: () => ({
    defaultAvatar: process.env.VUE_APP_DEFAULT_AVATAR,
    themes,
    languages: Object.values(languages),
    title: process.env.VUE_APP_TITLE,
    menuOpen: false
  }),
  computed: {
    ...mapGetters({
      authUser: "auth/authUser",
      profile: "auth/profile",
      theme: "theme",
      language: "language"
    }),
    $t() {
      return this.$translate(LANGUAGES_MAP, this.language.value);
    }
  },
  methods: {
    async logout() {
      const loggedOut = await this.$store.dispatch("auth/logout");
      if (loggedOut) {
        this.menuOpen = false;
        this.$router.replace("/login");
      }
    },
    changeTheme(theme) {
      this.$store.dispatch("changeTheme", theme);
    },
    changeLanguage(code) {
      this.menuOpen = false;
      this.$store.dispatch("changeLanguage", languages[code]);
    }
  }
};
</script>
