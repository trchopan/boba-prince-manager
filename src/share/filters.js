import Vue from "vue";

export const filters = {
  titleCase: str => {
    if (typeof str !== "string") {
      return "";
    }
    return str
      .toLowerCase()
      .split(" ")
      .map(function(word) {
        return word.replace(word[0], word[0].toUpperCase());
      })
      .join(" ");
  },
  format: (str, inputs) => {
    if (typeof str !== "string" || inputs.length === undefined) {
      return "";
    }
    let result = str;
    inputs.forEach(
      (input, index) => (result = result.replace(`{${index}}`, input))
    );
    return result;
  }
};

Object.keys(filters).forEach(key => Vue.filter(key, filters[key]));
