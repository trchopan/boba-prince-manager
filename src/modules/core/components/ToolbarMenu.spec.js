import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import { themes } from "@/modules/core/layout.models.js";
import { filters } from "@/share/filters.js";
import ToolbarMenu, { languagesMap } from "./ToolbarMenu.vue";
import { mockCustomElements } from "@/__mocks__/custom-elements.js";
import { docData } from "@/__mocks__/firebase-results";
import { languages } from "@/modules/core/layout.models.js";

const localVue = createLocalVue();
localVue.filter("titleCase", filters.titleCase);
localVue.use(Vuex);

describe("ToolbarMenu.vue", () => {
  let getters;
  let actions;
  let store;
  let wrapper;

  beforeEach(() => {
    getters = {
      "auth/profile": () => docData,
      "layout/theme": () => themes.light,
      "layout/language": () => languages.en
    };
    actions = {
      "layout/changeTheme": jest.fn(),
      "layout/changeLanguage": jest.fn(),
      "auth/logout": jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve(false))
        .mockImplementationOnce(() => Promise.resolve(true))
    };
    store = new Vuex.Store({
      getters,
      actions
    });
    wrapper = shallowMount(ToolbarMenu, {
      stubs: mockCustomElements,
      mocks: {
        $vuetify: { breakpoint: { mdAndUp: true } },
        $router: { replace: jest.fn() }
      },
      store,
      localVue
    });
  });

  it("logouts and handle error", async () => {
    await wrapper.vm.logout();
    expect(actions["auth/logout"]).toBeCalled();
    expect(wrapper.vm.$router.replace).not.toBeCalled();

    await wrapper.vm.logout();
    expect(wrapper.vm.$router.replace.mock.calls).toEqual([["/login"]]);
  });
  it("changes theme", () => {
    wrapper.vm.changeTheme(themes.light);
    expect(actions["layout/changeTheme"].mock.calls[0][1]).toEqual(
      themes.light
    );
  });
  it("changes language", () => {
    wrapper.vm.changeLanguage(languages.vi.code);
    expect(actions["layout/changeLanguage"].mock.calls[0][1]).toEqual(
      languages.vi.code
    );
  });
  it("renders profile correctly", () => {
    expect(wrapper.find("v-list-tile-title-stub").text()).toBe(
      docData.fullname
    );
    expect(wrapper.find("v-btn-stub").text()).toBe(languagesMap.logout.en);
    getters = {
      ...getters,
      "auth/profile": () => null
    };
    store = new Vuex.Store({
      getters,
      actions
    });
    wrapper = shallowMount(ToolbarMenu, {
      stubs: mockCustomElements,
      mocks: {
        $vuetify: { breakpoint: { mdAndUp: true } },
        $router: { replace: jest.fn() }
      },
      store,
      localVue
    });
    expect(wrapper.find("v-btn-stub").text()).toBe(languagesMap.login.en);
  });
  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
    getters = {
      ...getters,
      "auth/profile": () => null
    };
    store = new Vuex.Store({
      getters,
      actions
    });
    wrapper = shallowMount(ToolbarMenu, {
      stubs: mockCustomElements,
      mocks: {
        $vuetify: { breakpoint: { mdAndUp: true } },
        $router: { replace: jest.fn() }
      },
      store,
      localVue
    });
    expect(wrapper).toMatchSnapshot();
  });
});
