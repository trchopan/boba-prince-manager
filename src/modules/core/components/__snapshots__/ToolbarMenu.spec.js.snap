// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarMenu.vue matches snapshot 1`] = `
<v-menu-stub offset-y="" left="">
  <v-avatar-stub size="42"><img src="avatar" alt="avatar"></v-avatar-stub>
  <v-card-stub>
    <v-list-stub two-line="">
      <v-list-tile-stub avatar="" to="/profile">
        <v-list-tile-avatar-stub><img src="avatar" alt="big avatar"></v-list-tile-avatar-stub>
        <v-list-tile-content-stub>
          <v-list-tile-title-stub>Tester</v-list-tile-title-stub>
          <v-list-tile-sub-title-stub>tester@arkham.city</v-list-tile-sub-title-stub>
        </v-list-tile-content-stub>
      </v-list-tile-stub>
    </v-list-stub>
    <v-divider-stub></v-divider-stub>
    <v-list-stub>
      <v-list-tile-stub>
        <v-list-tile-content-stub>
          <v-select-stub value="[object Object]" items="[object Object],[object Object]" label="Select language"></v-select-stub>
        </v-list-tile-content-stub>
      </v-list-tile-stub>
      <v-list-tile-stub>
        <v-switch-stub label="Light out" input-value="Light" true-value="Dark" false-value="Light" class="pr-3 pt-2"></v-switch-stub>
      </v-list-tile-stub>
    </v-list-stub>
    <v-card-actions-stub>
      <v-spacer-stub></v-spacer-stub>
      <v-btn-stub outline="">Log out</v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-menu-stub>
`;

exports[`ToolbarMenu.vue matches snapshot 2`] = `
<v-menu-stub offset-y="" left="">
  <v-avatar-stub size="42"><img src="/images/bloody-smile.jpg" alt="avatar"></v-avatar-stub>
  <v-card-stub>
    <!---->
    <!---->
    <v-list-stub>
      <v-list-tile-stub>
        <v-list-tile-content-stub>
          <v-select-stub value="[object Object]" items="[object Object],[object Object]" label="Select language"></v-select-stub>
        </v-list-tile-content-stub>
      </v-list-tile-stub>
      <v-list-tile-stub>
        <v-switch-stub label="Light out" input-value="Light" true-value="Dark" false-value="Light" class="pr-3 pt-2"></v-switch-stub>
      </v-list-tile-stub>
    </v-list-stub>
    <v-card-actions-stub>
      <v-spacer-stub></v-spacer-stub>
      <v-btn-stub outline="" to="/login">Log in</v-btn-stub>
    </v-card-actions-stub>
  </v-card-stub>
</v-menu-stub>
`;
